File: September/23/addbinary.cs
// LeetCode 67: Add Binary
// Explanation:
// 1. Traverse from right to left with carry.
// 2. Build result string and reverse it at the end.
// Time Complexity: O(max(len(a), len(b)))
// Space Complexity: O(max(len(a), len(b)))

public class Solution {
    public string AddBinary(string a, string b) {
        int i = a.Length - 1, j = b.Length - 1, carry = 0;
        var sb = new System.Text.StringBuilder();
        
        while (i >= 0 || j >= 0 || carry > 0) {
            int total = carry;
            if (i >= 0) total += a[i--] - '0';
            if (j >= 0) total += b[j--] - '0';
            sb.Insert(0, (total % 2).ToString());
            carry = total / 2;
        }
        return sb.ToString();
    }
}

// Example usage:
// Solution sol = new Solution();
// Console.WriteLine(sol.AddBinary("11", "1"));      // "100"
// Console.WriteLine(sol.AddBinary("1010", "1011")); // "10101"

File: September/23/addbinary.py
# LeetCode 67: Add Binary
# Explanation:
# 1. Simulate binary addition from right to left with carry.
# 2. Append digits to result and reverse at the end.
# Time Complexity: O(max(len(a), len(b)))
# Space Complexity: O(max(len(a), len(b)))

class Solution:
    def addBinary(self, a: str, b: str) -> str:
        i, j, carry = len(a)-1, len(b)-1, 0
        res = []
        while i >= 0 or j >= 0 or carry:
            total = carry
            if i >= 0:
                total += int(a[i])
                i -= 1
            if j >= 0:
                total += int(b[j])
                j -= 1
            res.append(str(total % 2))
            carry = total // 2
        return "".join(reversed(res))
    
    def addBinaryBuiltIn(self, a: str, b: str) -> str:
        return bin(int(a, 2) + int(b, 2))[2:]
